torcontrol : avoid to set wrong outbound proxy and network settings when creating an inbound onion service.

Saibato opened the pull request:
While fiddling around with torcontrol and other tor proxy settings,
I noticed that the node tried to connect over the default 9050 port and ignored my settings. 

That was the case when the node try's to create a default hidden ephemeral tor service at startup.

So ,when we define our own proxy settings in conf, we don't want this to be skipped
by the default proxy creation that we setup while creating the
ephemeral hidden onion service.
Note; When -torcontrol and -proxy is defined in conf, we should *not*
override this with the default settings 127.0.0.1:9050.
Since the proxy ip and port might for good reasons have been set to other values.

Before this fix default settings could override conf and you could end
up not to be able to connect to .onion nodes or connect over the wrong proxy.

Also fixes #14722 some years ago noticed by qubenix 
EDIT:

While check more options I wondered what PR  #14425 is about?
And indeed the -onlynet flags are clearly disregarded  in respect to onions
since #7553. We could merge #14425 or adapt the tor.md doku to the actual
behavior.  Here my suggestion, with some *British* humor. :woman_shrugging:
My guess the consumption of the Catch22 movie, had influence when parts of
the option logic where implemented in the first place. 
 
edit saibato
The Tor doc change laanwj  mentioned later in review had moved to #20091 
This PR now also fixes #13378 and is replacement for the fix from wodry #14425

Saibato committed:
torconrol: avoid to set a wrong socks proxy when creating an onion service.

Before this fix default settings could override conf and you could end
up not be able to connect to .onion nodes or connect over the wrong proxy.

Side effect of this fix is that we no longer create the default
Tor socks outbound proxy, that we would have created by chance if an inbound
Tor service creation was successful.
Since the old version anyway would have leaked our ip, this fix
now dumps a warning that an outbound Tor proxy is not defined.

laanwj commented on src/torcontrol.cpp:
```diff
@@ -524,11 +524,15 @@ void TorController::auth_cb(TorControlConnection& _conn, const TorControlReply&
+            if (SetProxy(NET_ONION, addrOnion)) {
+                LogPrintf("tor: Default outbound tor .onion address proxy is set to 127.0.0.1:9050\n");
+                SetReachable(NET_ONION, true);
+            }
```
I think this should also log when `SetProxy` fails, as It means there is a bug in the software. After all it tries to set a hardcoded value which *should* never fail.

laanwj commented on src/torcontrol.cpp:
```diff
@@ -524,11 +524,15 @@ void TorController::auth_cb(TorControlConnection& _conn, const TorControlReply&
-            SetReachable(NET_ONION, true);
+            addrOnion = proxyType(resolved, gArgs.GetBoolArg("-proxyrandomize", true));
+            if (SetProxy(NET_ONION, addrOnion)) {
+                LogPrintf("tor: Default outbound tor .onion address proxy is set to 127.0.0.1:9050\n");
```
Please log to the right category:
```c++
LogPrint(BCLog::TOR, "tor: Default outbound tor .onion address proxy is set to 127.0.0.1:9050\n");
```

laanwj commented:
Concept ACK, good catch.

Concept ACK on the doc change as well. Though I'd suggest not refering to "nameproxy" as that's an internal implementation detail. I think the gist is:
```patch
+                       Note: Only the -proxy option will set the proxy used for DNS requests.
+                       With this option these will not route over Tor.
+                       So please use -proxy if you have privacy concerns with this.
```
This is the case because `-onion` is for multi-network nodes. You *only* want to use this option if you want your node to be able to connect to hidden services *as well*. Otherwise, always use `-proxy`.

BTW: as you found, the go-to document for configuring bitcoin over tor is `doc/tor.md`. It suggests using `-proxy` to route everything over tor. At some point it might make sense to refer to that instead of trying to cram everything in the option help.

> And indeed the -onlynet flags are clearly disregarded in respect to onions since #7553. 

Last time I checked,  `onlynet=onion` still works, in combination with `proxy=127.0.0.1:9050`, to only make outgoing connections to onion peers.

Saibato commented:
> Last time I checked, `onlynet=onion` still works, in combination with `proxy=127.0.0.1:9050`, to only make outgoing connections to onion peers.

thx, hmm, lets put it that way. Suppose a user selects -onlynet=ipv4 selects -listen  and  -proxy or none. I guess the least thing a human would expect in that case, is that the node would use a forgotten open existing tor port, to kindly out and inbound route onion traffic.
So i vote for an additional PR in line with #14425.

The more elleborate tor,md change when merged including your suggestions, at least warn now the user off this implications with the options use, but honestly, would you expect
a pure` bitcoind -listen -onlynet=ipv4 ` node would ever connect outbound *automatic* to tor ? 

Fact is, exactly that happens, a node that had an onion tor past and knows some onions, even later with a pure -onlynet=ipv4 conf, will open automatic outbound to those onions over tor, if a configed tor lingers forgotten or not  on that cruft.

Its not easy to craft nodes to do this in a one shot test setup, I try to  post here a bash script that does the steps, since that is a bit related to what we fix here, might be the right place?

jonatack commented on doc/tor.md:
```diff
@@ -29,8 +34,13 @@ outgoing connections, but more is possible.
-	                specified multiple times to allow multiple network types, e.g.
-	                ipv4, ipv6, or onion.
+	                specified multiple times to define a set of general allowed network
+	                types, e.g. ipv4,ipv6 or onion. If you use this option
```
s/,ipv6/, ipv6/

jonatack commented on src/torcontrol.cpp:
```diff
@@ -524,11 +524,15 @@ void TorController::auth_cb(TorControlConnection& _conn, const TorControlReply&
 
         // Now that we know Tor is running setup the proxy for onion addresses
```
nit: while here, can you please touch this up to add a missing comma and fix "setup", which is a noun but not a verb:

`s/running setup/running, set up/`

jonatack commented on src/torcontrol.cpp:
```diff
@@ -524,11 +524,15 @@ void TorController::auth_cb(TorControlConnection& _conn, const TorControlReply&
         // Now that we know Tor is running setup the proxy for onion addresses
         // if -onion isn't set to something else.
-        if (gArgs.GetArg("-onion", "") == "") {
+        // Or if we have already an outbound proxy defined. we skip this footgun gracefully.
```
```suggestion
        // Or if an outbound proxy is already defined, we can skip this footgun gracefully.
```

jonatack commented on src/torcontrol.cpp:
```diff
@@ -524,11 +524,15 @@ void TorController::auth_cb(TorControlConnection& _conn, const TorControlReply&
+            if (SetProxy(NET_ONION, addrOnion)) {
+                LogPrint(BCLog::TOR, "tor: Default outbound tor .onion address proxy is set to 127.0.0.1:9050\n");
+                SetReachable(NET_ONION, true);
+            } else LogPrint(BCLog::TOR, "tor: Error default outbound tor .onion address proxy for 127.0.0.1:9050 failed to create\n");
```
```suggestion
            } else {
                LogPrint(BCLog::TOR, "tor: Failed to create default outbound tor .onion address proxy for 127.0.0.1:9050\n");
            }
```

Saibato replied:
thx, had run it trough clang-format-diff and it also re-formated the net enum,  will come up with test and clang-diff fixup push that we that can autosquash ... , if you want to be co, i could commit the suggest as is, since i see it the same way, my while dev sarcasm comments must not be merged in the final version?

Saibato replied:
Also i think about, split the tor.md from this PR since wording yields always to intensive discussion.

jonatack reviewed (commented):
Concept ACK. Would you mind adding test coverage for the issue and code changes (perhaps in `test/functional/feature_proxy.py`)?

luke-jr commented:
Concept NACK, I think... If we're using `-torcontrol`, it makes sense to override `-proxy` (which is primarily for non-onion connections, and only used as a default for onions if no other is specified).

Perhaps a better alternative would be to disable `-torcontrol` (parameter interaction) if `-proxy` or `-onion` are set?

luke-jr reviewed (changes requested):
Concept NACK, I think... If we're using -torcontrol, it makes sense to override -proxy (which is primarily for non-onion connections, and only used as a default for onions if no other is specified).

Perhaps a better alternative would be to disable -torcontrol (parameter interaction) if -proxy or -onion are set?

Saibato commented:
> Concept NACK, I think... If we're using -torcontrol, it makes sense to override -proxy (which is primarily for non-onion connections, and only used as a default for onions if no other is specified).
> 
> Perhaps a better alternative would be to disable -torcontrol (parameter interaction) if -proxy or -onion are set?

I thought also about your suggestion, but dismissed that, since we have more problems that we should solve.

If you say pure -onion how to make sure the node does not dox the hidden service when the nameproxy is only selected by -proxy?  And -onlynet=onion had never worked for seeds since they use ipv4 connect and dns anyway in disregard of any conf and when you say -dns they resolve over tor and with 5 known correlated seeders on bootstrap you are doxed again.
And then when you say -dnseed=0,  after one hour or so when you least expect it the random conman selects one of the 3 only active fixed seed onions, oops doxed again. What is build here is the prefect catch22.  

Pls look also at #14425 and the patch for -onlynet i posted there,
The status quo should change anyway and i am open for suggestions, But when -torcontrol would override -proxy in general that was introduced when tor was enabled to make outbound sock5 onion connections, i see not really why we should mix inbound -torcontrol configurations with outbound proxy that cloud be a remote tor cruft  not in any way related to the local tor service port. 
     
Also #19043 -tortarget should be seen in that whole tor setup logic it make from a privacy perspective prefect sense to add such an option as you can easy forward over tor the hidden service from an other cruft, what makes things even more interesting if you want to have tx and mempool fill,  uncorrelated to your normal node operations and somewhat quasi on the fly could enable this,.

Cloud you outline how you would see the tor implementation work?
Want we have now is not in line with the documentation, so we have to change anyway.
And since our implementation with the keys stored permanent is as ephemeral as the holy night.
I would like to mention here that we have with the Tor browser a stress free on click way to generate new hidden true ephemeral services on the fly i.e. over 9150/51 for posting as temp 3D code for short inbound. I guess something you would like to have from an other PR of yours. What is a good idea and nice to have..
.
so hmm.. what now?.

Saibato commented:
changed hidden to onion and rebased https://github.com/bitcoin/bitcoin/pull/19358/commits/468c60c5ed908ca8029db241b75aa8f893866c74 https://github.com/bitcoin/bitcoin/pull/19358/commits/654dc04b67a44199352a1b25494c9d617cd090bc

qubenix commented:
Tested 654dc04b67a44199352a1b25494c9d617cd090bc, does indeed fix #14722. This is obviously a better fix than #14729.

fanquake commented:
> This is obviously a better fix than #14729.

Thanks for testing @qubenix. I'll close #14729.

fanquake commented:
@Saibato note that you also need to [remove any @ mentions from your commit messages](https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md#committing-patches), 654dc04b67a44199352a1b25494c9d617cd090bc in this case. Otherwise those users will likely be spammed with mentions on GitHub.

Saibato commented:
Rebased  and changed the commit messages of c40a121 and   https://github.com/bitcoin/bitcoin/pull/19358/commits/efacd5f4f921e46652c4ef0ae83cadd782395793
like @fanquake suggested.

Btw; the more "correct" PR description should be.
**torcontrol :do not override silent behind the back of the user with default proxy settings in hidden service, if the user has chosen other settings according with the wording in the docs**

hebasto reviewed (commented):
Concept ACK on making config options interaction more clear.

Approach NACK. Only establishing an onion service for _inbound_ Tor connections should be the sole responsibility of the `TorController` class. Keeping in its implementation the code which deals with a proxy settings for _outbound_ Tor connections looks like a bad design, that leads to behavior like [that](https://github.com/bitcoin/bitcoin/issues/14722#issuecomment-438639561).

Saibato commented:
> t leads to behavior like ...

@hebasto  U right, only at first glance it seams the right and only possible place since you can not know in **init** if the async call fails or succeeds
Might also be the reason no one saw that for a long time at first glance when reviewing ie. `src/init.cpp` that setting the proxy would be overruled later.
its a real mess and we should assume shit happens.
And now we fix it. 

I will push an alternative that try's to solves that partly in` `init.cpp`. by assuming a partly successful call and place a comment there that we set the final proxy when we return the socket and announce the address.

Saibato commented:
Rebased https://github.com/bitcoin/bitcoin/pull/19358/commits/2b9572bf298ca9d4307ec7fe3f5ab8aba4f15e7f and renamed the PR and changed the pr commit message and stripped the doc https://github.com/bitcoin/bitcoin/commit/efacd5f4f921e46652c4ef0ae83cadd782395793 from the PR in favor of a new tor doc PR

Saibato commented:
Rebased fixup commit https://github.com/bitcoin/bitcoin/commit/2b9572bf298ca9d4307ec7fe3f5ab8aba4f15e7f we now init the outbound proxy in` init.cpp` 
Addressed also review comment from @hebasto 
It leaves torcontrol to solely do inbound stuff and init;s the outbound stuff in ` init.cpp`


Its also now a repclament for #14425 and fixes #13378

vasild commented on src/torcontrol.cpp:
```diff
@@ -527,15 +527,6 @@ void TorController::auth_cb(TorControlConnection& _conn, const TorControlReply&
     if (reply.code == 250) {
         LogPrint(BCLog::TOR, "tor: Authentication successful\n");
 
-        // Now that we know Tor is running setup the proxy for onion addresses
-        // if -onion isn't set to something else.
-        if (gArgs.GetArg("-onion", "") == "") {
-            CService resolved(LookupNumeric("127.0.0.1", 9050));
-            proxyType addrOnion = proxyType(resolved, true);
-            SetProxy(NET_ONION, addrOnion);
-            SetReachable(NET_ONION, true);
-        }
-
```
Isn't this removal sufficient to fix https://github.com/bitcoin/bitcoin/issues/14722?

Saibato replied:
in some sense yes, i would also feel better to just remove this since something like this was anyway somehow without the users explicit consent.
But i guess the main concern was to leak ip, if there is no proxy, but the implementation we had now for years was sub optimal at best.

If we can find there common ground to have by default no outbound Tor socks proxy, i am fine with this to just remove this. :-). 
On the other hand from privacy point it might be good to only generate inbound onion services, if there is a Tor proxy set, to not leak ip.
And maybe do a hint comment to the users to encourage them to define at least -proxy= and write some hints in the docs.

And then do an extra PR for the -onlynet=onion.

Saibato replied:
@vasild  thx for also looking at this. u might also notice that since -listenonion is the default, i am thinking right now, why not just create the service if we have a proxy in first place?

Bitcoin's net is now big enough and we no longer need to create random onions to grow the net, so if -listenonion is removed as default we would have a bit more privacy and a clear torcontrol api.

vasild replied:
Correct me if I am wrong, if only the above code snippet is removed (no other changes), then the behavior will be:

`-proxy=` provided | `-onion=` provided | outcome
---|---|---
no | no | outgoing tor connections are not possible
no | yes | outgoing tor connections will be made via the proxy provided in `-onion=`
yes | no | outgoing tor connections will be made via the proxy provided in`-proxy=` (`-onion=` defaults to `-proxy=`)
yes | yes | outgoing tor connections will be made via the proxy provided in `-onion=`

regardless of whether `-torcontrol=` is provided or not. Which to me makes perfect sense. Users that want to avoid clearnet outgoing connections should set `-proxy=` to the tor socks5 proxy.

I think fixing `-onlynet=` is a separate issue and warrants a separate PR.

Saibato replied:
Makes sense and i addressed ur suggestion.

We now will not create the default outbound proxy and fix by that also the issue and dump a reminder in log  that the user might leak his ip when no proxy is defined and the service call succeeds later somehow on purpose or i.e. when he enables Tor and the node runs already with listenonion, so at least better than the old footgun.

And since we now do not override -onlynet in torcontrol even in that regard better than what we had since years.

wodry replied:
> Correct me if I am wrong, if only the above code snippet is removed (no other changes), then the behavior will be:

Thank you for the work of analyzing the behavior, @vasild and creating the overview table, this is very good.

Please see laanwj's destinction of when using names "Tor" and when using "Onion" here:
https://github.com/bitcoin/bitcoin/pull/13575#issuecomment-402427844

As I understand, the `onion=` option is only about "onion" connections (that is, connections to .onion Onion services (aka deprecated name Hidden service).

While the `proxy=` setting, when set to a Tor node, forwards IPv4 (non .onion) requests to the Tor network, when they go out to the clearnet from a Tor exit node. As I underdstand, this would be named "tor connection".

So, e.g. if `proxy` is unset and `onion` is set, then there will be Onion (and maybe IPv4/IPv6) connections, but no Tor connections.

And if `proxy` is set and `onion` is unset, then there will only be Tor connections connections (both via exit node to clearnet and to Onion services).

Would be an experiment to set both `proxy` and `onion` to different tor servers and check if really IPv4 connections go through the proxy-Tor server, and .onion connections go through the onion-Tor server. But I think this is not really relevant as I can not think of a use case of having two Tor servers to separate tor and onion connections.

So I think the definition would be:
- Tor connection: Connection to IPv4 (and IPv6?) clearnet through Tor network tunnel, and optional to .onion services.
- Onion connection: Connection to .onion service.

Saibato committed:
init: followup up fix of -onlynet settings.

While the onlynet=onion problem is fixed by not creating and outbound proxy
in torcontrol.
We now with this fix respect also the possible fine grained ipv4/ipv6 setting
combinations.

vasild commented on src/init.cpp:
```diff
@@ -1975,6 +1975,10 @@ bool AppInitMain(const util::Ref& context, NodeContext& node, interfaces::BlockA
         }
         StartTorControl(bind_addr);
+        proxyType addrOnion;
+        if (gArgs.GetArg("-onion", "") == "" && !GetProxy(NET_ONION, addrOnion) ) {
+            LogPrintf("Tor: reminder: listenonion is not disabled and an outbound Tor proxy is not defined, that could leak your ip\n");
+        }
```
I think just `!GetProxy(NET_ONION, addrOnion)` is enough. If the second expression is `true` then for sure the first one is `true` too.

But more importantly, I think this log message is unnecessary - the fact that a node is running a tor onion service does not mean they want to hide their IP address from the world. I guess `-onlynet=onion` implies that.

And even if we want to log a message - the above logic is flawed because we would not log it if `-onion=` is provided but `-proxy=` is not and thus the node would make outgoing IPv4 connections directly (revealing its IP address).

Saibato replied:
In my view Its the other way around , the first might be false like u noticed and the use wants to leak his ip with -onion? 
Since -onion is always without nameproxy and the user must know that that leaks anyway.
So i thought we should not warn him when he wish for that.

To have no -proxy set is in my view more concerning.
At least with deleting the nameproxy leaking  version by default we had until now
we overall are better off with ur suggestion to delete it at all. 

There is no clean solution, when we say -listenonion by default.
even if we use  `IsReachable(NET_ONION) `we could not skip all cases, since all nets are on by default.

Like i said in an other PR its a mess and we should fix it before the transition to v3.
Not even thinking about seeding and the -onlynet flags, its doxing mess wherever u dig deep.
Its almost often counter intuitive crafted and the reverse of what u define with some *privacy* options.

Saibato replied:
@vasild could u imagine that we define a flag -privacynode=maximum,medium,low.
That is by default on maximum?

i.e. maximum no listenonion if no proxy with nameproxy no seeding no dns no uunp onlynet respected everywhere just v3 no persistent keystore bind only to unrouteable addr if Tor etc.
or low just as is.

That would really make things more easy to craft in the different modules core will split anyway, by not annoying the user with warnings and a clear guideline for devs .to
follow when obeying those setting.

Saibato commented:
Rebased addressed review and added the detrimental followup -onlynet fixup.commit https://github.com/bitcoin/bitcoin/pull/19358/commits/e9452a1a8f673de1b310e918ae80c31b7be6f171

Saibato commented:
Rebased https://github.com/bitcoin/bitcoin/commit/901ac08bdb400aece7a8d910da977acea65fa6af fixed the -onlynet settings for proxy use  ready to review.

Willtech commented:
Please note this should resolve Issue #12641 when validating. -DA.

Saibato commented:
> Please note this should possibly resolve Issue #12641 when validating? -DA.

I guess what u observe is that Tor can for sure over ip4/6 Tor exit nodes connect to nodes in the ip4/6 clearnet. if the address given to the socks5 proxy is not an onion address like from your seed.dat or address learned from other nodes.. Did u expect with -proxy just Tor onion address outbound connections?
If so, please set also -onlynet=onion, pls ping me when even with this you still see outbound non onion connections..
I observed so far only the other way around, lingered forgotten onions in peers,dat would connect outbound if -onlynet=ipv4 or 6  is set. ooops.

Willtech commented:
Good Afternoon,
Reply via email.
> I guess what u observe is that Tor can for sure over ip4/6 Tor exit
> nodes connect to nodes in the ip4/6 clearnet. 
Well, yes if it is not set `onlynet=onion`
> Did u expect with -proxy just Tor onion address outbound connections?
Both onion and proxy are set to the tor proxy.
> -onlynet=onion
onlynet=onion is fine and connections to pure onion addresses works,
that is not the problem, the problem is not respecting the proxy
setting.
Regards,The AustralianLORD HIS EXCELLENCY JAMES HRMH (& HMRH)of Hougun
Manor & Glencoe & British EmpireMR. Damian A. James WilliamsonWills
et al.
 Willtechwww.willtech.com.auwww.go-overt.comand other
projects earn.com/willtechlinkedin.com/in/damianwilliamson

m. 0487135719f. 61261470192

--On Thu, 2020-10-15 at 23:02 -0700, Saibato wrote:
> > Please note this should possibly resolve Issue #12641 when
> > validating? -DA.
> 
> I guess what u observe is that Tor can for sure over ip4/6 Tor exit
> nodes connect to nodes in the ip4/6 clearnet. if the address given to
> the socks5 proxy is not an onion address like from your seed.dat or
> address learned from other nodes.. Did u expect with -proxy just Tor
> onion address outbound connections?
> 
> If so, please set also -onlynet=onion, pls ping me when even with
> this you still see outbound non onion connections..
> 
> I observed so far only the other way around, lingered forgotten
> onions in peers,dat would connect outbound if -onlynet=ipv4 or 6  is
> set. ooops.
> 
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub, or unsubscribe.
> 
> [
> {
> "@context": "http://schema.org",
> "@type": "EmailMessage",
> "potentialAction": {
> "@type": "ViewAction",
> "target": "
> https://github.com/bitcoin/bitcoin/pull/19358#issuecomment-709815857"
> ,
> "url": "
> https://github.com/bitcoin/bitcoin/pull/19358#issuecomment-709815857"
> ,
> "name": "View Pull Request"
> },
> "description": "View this Pull Request on GitHub",
> "publisher": {
> "@type": "Organization",
> "name": "GitHub",
> "url": "https://github.com"
> }
> }
> ]
